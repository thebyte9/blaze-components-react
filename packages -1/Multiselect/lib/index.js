"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var checkboxes_1 = __importDefault(require("@blaze-react/checkboxes"));
var chips_1 = __importDefault(require("@blaze-react/chips"));
var input_1 = __importDefault(require("@blaze-react/input"));
var utils_1 = __importDefault(require("@blaze-react/utils"));
var lodash_differencewith_1 = __importDefault(require("lodash.differencewith"));
var lodash_isequal_1 = __importDefault(require("lodash.isequal"));
var lodash_unionby_1 = __importDefault(require("lodash.unionby"));
var react_1 = __importStar(require("react"));
var MultiSelect = function (_a) {
    var _b = _a.data, data = _b.data, keys = _b.filterBy, keyValue = _b.keyValue, identification = _b.identification, _c = _a.utils, ErrorMessage = _c.ErrorMessage, uniqueId = _c.uniqueId, validationMessage = _a.validationMessage, notFoundMessage = _a.notFoundMessage, getSelected = _a.getSelected, placeholder = _a.placeholder, children = _a.children, onChange = _a.onChange, error = _a.error, name = _a.name, selected = _a.selected;
    var _d = react_1.useState([]), dataCopy = _d[0], setDataCopy = _d[1];
    var _e = react_1.useState(false), show = _e[0], setShow = _e[1];
    react_1.useEffect(function () {
        var shouldUpdate = lodash_differencewith_1["default"](dataCopy, data, lodash_isequal_1["default"]).length ||
            lodash_differencewith_1["default"](data, dataCopy, lodash_isequal_1["default"]).length;
        var elementsWithSelected = lodash_unionby_1["default"](selected, data, "id");
        if (!dataCopy || shouldUpdate) {
            setDataCopy(elementsWithSelected.map(function (element) {
                element.show = true;
                return element;
            }));
        }
    }, [data]);
    var setStatus = function (obj, status) {
        return Object.assign({}, obj, { show: status });
    };
    var handleInputChange = function (_a) {
        var event = _a.event, value = _a.value;
        var parsedDataCopy = parseDataCopy(value);
        if (onChange) {
            onChange({ event: event, value: value, name: name });
        }
        setDataCopy(parsedDataCopy);
    };
    var parseDataCopy = function (value) {
        return dataCopy.map(function (copy) {
            return setStatus(copy, !!keys.some(function (key) {
                var copyKey = copy[key].toString().toLowerCase();
                return copyKey.includes(value.toLowerCase());
            }));
        });
    };
    var updateData = function (element, value) {
        var newDataCopy = dataCopy.slice();
        newDataCopy[element].checked = value;
        setDataCopy(newDataCopy);
    };
    var handleKeyDown = function (_a) {
        var keyName = _a.key, value = _a.target.value;
        if (keyName === "Enter") {
            var parsedDataCopy = parseDataCopy(value);
            var elementToAdd = parsedDataCopy.findIndex(function (parsedData) { return parsedData.show; });
            updateData(elementToAdd, true);
        }
    };
    var handleCheckBoxChange = function (_a) {
        var value = _a.value, localData = _a.data;
        setDataCopy(localData);
        getSelected({
            event: {
                target: {
                    name: name,
                    value: value
                }
            }
        });
    };
    var parseCheckBoxOptions = function (elements) {
        return elements.map(function (element) { return (__assign({}, element, { label: element[keyValue] })); });
    };
    var handleDelete = function (id) {
        var elementToDelete = dataCopy.findIndex(function (_a) {
            var itemId = _a.id;
            return itemId === id;
        });
        updateData(elementToDelete, false);
    };
    var handleClearAll = function () {
        var formatedElements = dataCopy.map(function (item) {
            item.checked = false;
            return item;
        });
        setDataCopy(formatedElements);
        setShow(false);
    };
    var handleFocus = function () { return setShow(true); };
    var matchQuery = !!dataCopy.filter(function (item) { return item.show; })
        .length;
    return (react_1["default"].createElement(react_1["default"].Fragment, null,
        react_1["default"].createElement("div", { className: "multiselect" },
            dataCopy
                .filter(function (item) { return item.checked; })
                .map(function (selectedValue) { return (react_1["default"].createElement(chips_1["default"], { modifiers: [
                    chips_1["default"].availableModifiers.parent.deletable,
                    chips_1["default"].availableModifiers.parent.small
                ], onDelete: function () { return handleDelete(selectedValue[identification]); }, action: function () { return handleDelete(selectedValue[identification]); }, key: uniqueId(selectedValue) },
                react_1["default"].createElement(chips_1["default"].Label, null, selectedValue[keyValue]),
                react_1["default"].createElement(chips_1["default"].Icon, { modifier: chips_1["default"].availableModifiers.icon["delete"] },
                    react_1["default"].createElement("i", { className: "material-icons" }, "clear")))); }),
            children,
            react_1["default"].createElement(input_1["default"], { placeholder: placeholder, onChange: handleInputChange, onKeyDown: handleKeyDown, onFocus: handleFocus, className: "multiselect__input" }),
            show && (react_1["default"].createElement("div", { className: "multiselect__dropdown" },
                error && react_1["default"].createElement(ErrorMessage, { message: validationMessage }),
                !matchQuery && react_1["default"].createElement("p", null, notFoundMessage),
                react_1["default"].createElement(checkboxes_1["default"], { options: parseCheckBoxOptions(dataCopy), onChange: handleCheckBoxChange }))),
            react_1["default"].createElement("span", { className: "multiselect__clear", onClick: handleClearAll },
                react_1["default"].createElement("i", { className: "material-icons" }, "clear")))));
};
MultiSelect.defaultProps = {
    children: "",
    error: false,
    getSelected: function () { return void 0; },
    notFoundMessage: "No records available",
    onChange: function (arg) {
        return arg;
    },
    placeholder: "Search...",
    validationMessage: "This field is required"
};
exports["default"] = utils_1["default"](MultiSelect);
