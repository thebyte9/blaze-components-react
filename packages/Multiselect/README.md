## Description

MultiSelect is a component that allows you to select multiple items with check boxes.
It is useful for labeling, contact lists, country selectors, etc.

## Usage

```js
const data = {
  identification: "id",
  keyValue: "name",
  filterBy: ["name", "description"],
  data: [
    {
      show: true,
      checked: true,
      id: 1,
      name: "Blaze",
      description: "Lorem ipsum dolor."
    },
    {
      show: true,
      checked: false,
      id: 2,
      name: "KP",
      description: "Aliquam tincidunt."
    },
    {
      show: true,
      checked: true,
      id: 3,
      name: "Pulser",
      description: "Vestibulum auctor."
    }
  ]
};

<Multiselect data={data} />;
```

## API

##### Multiselect can receive a number of `props` as follow:

| NAME         |         TYPE          | DEFAULT  |
| :----------- | :-------------------: | :------: |
| data         |        Object         |    {}    |
| limit        |        Number         |    0     |
| placeholder  |        String         | 'Search' |
| toggleButton |        String         |  empty   |
| selected     |       Function        | () => {} |
| children     | single/array of nodes |  empty   |
