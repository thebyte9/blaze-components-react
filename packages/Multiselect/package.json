{
  "name": "@blaze-react/multiselect",
  "version": "0.8.0-alpha.63",
  "description": "",
  "main": "./dist/main.js",
  "scripts": {
    "build": "rm -Rf ./dist && NODE_ENV=production webpack --config webpack.config.js",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "eslint '*/**/*.{ts,tsx}' --ignore-path ../../.eslintignore --fix",
    "test": "jest --passWithNoTests",
    "test:coverage": "jest --coverage --passWithNoTests",
    "prepublishOnly": "yarn build",
    "bundle-analyser": "node ../../node_modules/webpack-bundle-analyzer/lib/bin/analyzer.js ./stats/stats.json"
  },
  "keywords": [],
  "author": "byte9",
  "license": "GPL-3.0",
  "devDependencies": {
    "@babel/cli": "^7.13.14",
    "@babel/core": "^7.13.14",
    "@types/lodash.unionby": "^4.8.6",
    "lodash.differencewith": "^4.5.0",
    "lodash.isequal": "^4.5.0",
    "lodash.unionby": "^4.8.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-tiny-virtual-list": "^2.2.0",
    "webpack": "^5.28.0",
    "webpack-cli": "^4.6.0"
  },
  "peerDependencies": {
    "@blaze-react/checkboxes": "^0.8.0-alpha.58",
    "@blaze-react/utils": "^0.8.0-alpha.58",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-tiny-virtual-list": "^2.2.0"
  },
  "gitHead": "d9cac6fc28dad1d0794718aac09d292019446698"
}