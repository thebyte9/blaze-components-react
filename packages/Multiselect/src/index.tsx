import Checkboxes from "@blaze-react/checkboxes";
import Input from "@blaze-react/input";
import differenceWith from "lodash.differencewith";
import isEqual from "lodash.isequal";
import React, { useEffect, useState } from "react";
import uuidv1 from "uuid/v1";

interface IMultiSelectProps {
  data: {
    keyValue: string;
    filterBy: any[];
    data: object[];
  };
  selected: (...args: any[]) => any;
  placeholder?: string;
  children?: any;
  onChange?: (arg: { event: Event; value: string }) => void;
}
const MultiSelect: React.SFC<IMultiSelectProps> = ({
  data: { data, filterBy: keys, keyValue },
  selected: getSelected,
  placeholder,
  children,
  onChange
}): JSX.Element => {
  const [selected, setSelected] = useState<any[]>([]);
  const [dataCopy, setDataCopy] = useState<object[]>(data);

  useEffect(() => {
    const shouldUpdate =
      differenceWith(dataCopy, data, isEqual).length ||
      differenceWith(data, dataCopy, isEqual).length;
    if (shouldUpdate) {
      setDataCopy(data);
    }
  }, [data]);

  const setStatus = (obj: any, status: any): object =>
    Object.assign({}, obj, { show: status });

  const handleInputChange = ({
    event,
    value
  }: {
    event: Event;
    value: string;
  }) => {
    const parsedDataCopy = dataCopy.map((copy: any) =>
      setStatus(
        copy,
        !!keys.some(key => {
          const copyKey = copy[key].toLowerCase();
          return copyKey.includes(value.toLowerCase());
        })
      )
    );

    if (onChange) {
      onChange({ event, value });
    }
    setDataCopy(parsedDataCopy);
  };

  const handleCheckBoxChange = ({
    value,
    data: localData
  }: {
    value: any;
    data: any;
  }) => {
    setSelected(value);
    setDataCopy(localData);
    getSelected({event: {
      target: {
        value: localData
      }
    }});
  };

  const parseCheckBoxOptions = (elements: object[]): object[] =>
    elements.map(
      (copiedData: any): object => ({
        ...copiedData,
        label: copiedData[keyValue]
      })
    );

  return (
    <>
      {selected.map(
        (selectedValue): JSX.Element => (
          <div key={uuidv1()}>{selectedValue[keyValue]}</div>
        )
      )}
      {children}
      <Input placeholder={placeholder} onChange={handleInputChange} />
      <Checkboxes
        options={parseCheckBoxOptions(dataCopy)}
        onChange={handleCheckBoxChange}
      />
    </>
  );
};
MultiSelect.defaultProps = {
  children: "",
  onChange: (arg: { event: Event; value: string }) => {
    return arg;
  },
  placeholder: "Search",
  selected: () => {
    return;
  }
};
export default MultiSelect;
