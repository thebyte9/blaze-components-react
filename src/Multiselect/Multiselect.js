import React, { useState, Fragment } from 'react';
import PropTypes from 'prop-types';
import uuidv1 from 'uuid/v1';

import Input from '../Input';
import Checkboxes from '../Checkboxes';

const MultiSelect = ({
  data: { data, filterBy: keys, keyValue },
  selected: getSelected,
  placeholder,
  children
}) => {
  const [selected, setSelected] = useState([]);
  const [dataCopy, setDataCopy] = useState(data);

  const setStatus = (obj, status) => Object.assign({}, obj, { show: status });

  const handleKeyUp = (event) => {
    const { target: { value } } = event;

    const _dataCopy = dataCopy.map((copy) => {
      let _copy = setStatus(copy, false);

      keys.forEach((_key) => {
        const match = copy[_key].toLowerCase().includes(value.toLowerCase());
        if (match) _copy = setStatus(copy, true);
      });

      return _copy;
    });

    setDataCopy(_dataCopy);
  };

  const handleChange = ({ checked, data: _data }) => {
    setSelected(checked);
    setDataCopy(_data);
    getSelected(_data);
  };

  return (
    <Fragment>

      {
        selected.map(_selected => <div key={uuidv1()}>{_selected[keyValue]}</div>)
      }

      {children}

      <Input placeholder={placeholder} onKeyUp={handleKeyUp} />
      {
        <Checkboxes
            options={dataCopy.map(_dataCopy => (
              Object.assign({}, _dataCopy, { label: _dataCopy[keyValue] })))}
            onChange={handleChange}
            withEffect
            />
      }
    </Fragment>
  );
};

MultiSelect.propTypes = {
  data: PropTypes.object,
  selected: PropTypes.func,
  placeholder: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};

MultiSelect.defaultProps = {
  data: {},
  selected: () => {},
  placeholder: 'Search',
  children: ''
};

export default MultiSelect;

