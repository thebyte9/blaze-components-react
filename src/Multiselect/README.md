
## Description

MultiSelect is a component that allows you to select multiple items with check boxes. 
It is useful for labeling, contact lists, country selectors, etc.

## Usage

```js
const data = {
  keyValue: 'name',
  filterBy: ['name', 'description'],
  data: [
    {
        id: 1,
        name: 'Blaze',
        description: 'Lorem ipsum dolor.'
    },
    {
        id: 2,
        name: 'KP',
        description: 'Aliquam tincidunt.'
    },
    {
        id: 3,
        name: 'Pulser',
        description: 'Vestibulum auctor.'
    }
  ]
};

<Multiselect data={data} selected={(result) => {}}/>
```

## API

##### Multiselect can receive a number of `props` as follow:
clea
| NAME   | TYPE | DEFAULT | 
| :---  | :---:  | :---: | 
| data | Object | {} |
| placeholder | String | 'Search' |
| toggleButton | String | empty | 
| selected | Function | () => {} | 
| children | single/array of nodes | empty | 





