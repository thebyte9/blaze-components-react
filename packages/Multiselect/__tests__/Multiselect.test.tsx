import React from 'react';
import expect from 'expect';
import { shallow, mount } from 'enzyme';
import Multiselect from '../src';

const data = {
  keyValue: 'name',
  filterBy: ['name'],
  data: [
    {
      id: 1,
      name: 'Project1'
    },
    {
      id: 2,
      name: 'Project2'
    },
    {
      id: 3,
      name: 'Project3'
    }
  ]
};

const MultiselectComponent = <Multiselect data={data} selected={() => {}} />;

describe('DragableArea component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const wrapper = shallow(MultiselectComponent);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test('should select first option', () => {
    const wrapper = mount(MultiselectComponent);

    wrapper
      .find('.form-field--checkbox')
      .at(0)
      .simulate('click');

    expect(
      wrapper
        .find('div')
        .at(0)
        .text()
    ).toContain('Project1');
  });

  test('should allow to filter', () => {
    const wrapper = mount(MultiselectComponent);

    wrapper
      .find('input')
      .at(0)
      .simulate('keyUp', { target: { value: 'Project2' } });
    expect(
      wrapper
        .find('.form-field--checkbox')
        .at(0)
        .text()
    ).toContain('Project2');
  });
});
