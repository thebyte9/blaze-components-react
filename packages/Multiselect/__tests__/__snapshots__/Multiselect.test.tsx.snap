// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiselect component should be defined and renders correctly (snapshot) 1`] = `
<MultiSelect
  data={
    Object {
      "data": Array [
        Object {
          "id": 1,
          "name": "Blaze 1",
        },
        Object {
          "id": 2,
          "name": "Blaze 2",
        },
        Object {
          "id": 3,
          "name": "Blaze 3",
        },
      ],
      "filterBy": Array [
        "name",
      ],
      "identification": "id",
      "keyValue": "name",
    }
  }
  error={false}
  getSelected={[Function]}
  notFoundMessage="No records available"
  onChange={[Function]}
  placeholder="Search..."
  selected={
    Array [
      Object {
        "checked": true,
        "description": "Lorem ipsum dolor.",
        "id": 1,
        "name": "Blaze 1",
      },
    ]
  }
  utils={
    Object {
      "ErrorMessage": [Function],
      "classNames": [Function],
      "uniqueId": [Function],
    }
  }
  validationMessage="This field is required"
/>
`;
